File: cloud/blockstore/public/sdk/go/client/request.go
package client

import (
	"fmt"
	"math/rand"
	"strings"

	"github.com/ydb-platform/nbs/cloud/blockstore/public/api/protos"
	storage_protos "github.com/ydb-platform/nbs/cloud/storage/core/protos"
)

////////////////////////////////////////////////////////////////////////////////

type request interface {
	GetHeaders() *protos.THeaders
	String() string
}

////////////////////////////////////////////////////////////////////////////////

func nextRequestId() uint64 {
	var requestId uint64 = 0
	for requestId == 0 {
		requestId = rand.Uint64()
	}
	return requestId
}

func requestName(req request) string {
	name := underlyingTypeName(req)
	if strings.HasPrefix(name, "T") && strings.HasSuffix(name, "Request") {
		// strip "T" prefix and "Request" suffix from type name
		name = name[1 : len(name)-7]
	}
	return name
}

func requestSize(req request) uint32 {
	if readReq, ok := req.(*protos.TReadBlocksRequest); ok {
		return readReq.GetBlocksCount()
	}

	if writeReq, ok := req.(*protos.TWriteBlocksRequest); ok {
		return uint32(len(writeReq.Blocks.Buffers))
	}

	if zeroReq, ok := req.(*protos.TZeroBlocksRequest); ok {
		return zeroReq.GetBlocksCount()
	}

	return 0
}

func requestLogLevel(req request) LogLevel {
	switch req.(type) {
	case *protos.TReadBlocksRequest:
		return LOG_DEBUG
	case *protos.TWriteBlocksRequest:
		return LOG_DEBUG
	case *protos.TZeroBlocksRequest:
		return LOG_DEBUG
	default:
		return LOG_INFO
	}
}

func requestDetails(req request) string {
	if readReq, ok := req.(*protos.TReadBlocksRequest); ok {
		return fmt.Sprintf(
			" (offset: %d, blocks count: %d)",
			readReq.StartIndex,
			readReq.BlocksCount)
	}

	if writeReq, ok := req.(*protos.TWriteBlocksRequest); ok {
		return fmt.Sprintf(
			" (offset: %d, buffers count: %d)",
			writeReq.StartIndex,
			len(writeReq.Blocks.Buffers))
	}

	if zeroReq, ok := req.(*protos.TZeroBlocksRequest); ok {
		return fmt.Sprintf(
			" (offset: %d, blocks count: %d)",
			zeroReq.StartIndex,
			zeroReq.BlocksCount)
	}

	if createVolumeReq, ok := req.(*protos.TCreateVolumeRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, project id: %s, block size: %d, blocks count: %d, channels count: %d, storage media kind %s, folder id: %s, cloud id: %s)",
			createVolumeReq.DiskId,
			createVolumeReq.ProjectId,
			createVolumeReq.BlockSize,
			createVolumeReq.BlocksCount,
			createVolumeReq.ChannelsCount,
			storage_protos.EStorageMediaKind_name[int32(createVolumeReq.StorageMediaKind)],
			createVolumeReq.FolderId,
			createVolumeReq.CloudId)
	}

	if destroyVolumeReq, ok := req.(*protos.TDestroyVolumeRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s)",
			destroyVolumeReq.DiskId)
	}

	if resizeVolumeReq, ok := req.(*protos.TResizeVolumeRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, blocks count: %d, channels count: %d)",
			resizeVolumeReq.DiskId,
			resizeVolumeReq.BlocksCount,
			resizeVolumeReq.ChannelsCount)
	}

	if alterVolumeReq, ok := req.(*protos.TAlterVolumeRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, project id: %s, folder id: %s, cloud id: %s)",
			alterVolumeReq.DiskId,
			alterVolumeReq.ProjectId,
			alterVolumeReq.FolderId,
			alterVolumeReq.CloudId)
	}

	if assignVolumeReq, ok := req.(*protos.TAssignVolumeRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, instance id: %s, host: %s)",
			assignVolumeReq.DiskId,
			assignVolumeReq.InstanceId,
			assignVolumeReq.Host)
	}

	if mountVolumeReq, ok := req.(*protos.TMountVolumeRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, instance id: %s, volume access mode: %s, volume mount mode: %s, ipc type: %s, mount seq number: %d)",
			mountVolumeReq.DiskId,
			mountVolumeReq.InstanceId,
			protos.EVolumeAccessMode_name[int32(mountVolumeReq.VolumeAccessMode)],
			protos.EVolumeMountMode_name[int32(mountVolumeReq.VolumeMountMode)],
			protos.EClientIpcType_name[int32(mountVolumeReq.IpcType)],
			mountVolumeReq.MountSeqNumber)
	}

	if unmountVolumeReq, ok := req.(*protos.TUnmountVolumeRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, instance id: %s, session id: %s)",
			unmountVolumeReq.DiskId,
			unmountVolumeReq.InstanceId,
			unmountVolumeReq.SessionId)
	}

	if createCheckpointReq, ok := req.(*protos.TCreateCheckpointRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, checkpoint id: %s)",
			createCheckpointReq.DiskId,
			createCheckpointReq.CheckpointId)
	}

	if deleteCheckpointReq, ok := req.(*protos.TDeleteCheckpointRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, checkpoint id: %s)",
			deleteCheckpointReq.DiskId,
			deleteCheckpointReq.CheckpointId)
	}

	if getChangedBlockReq, ok := req.(*protos.TGetChangedBlocksRequest); ok {
		return fmt.Sprintf(
			" (disk id: %s, start index: %d, blocks count: %d, low checkpoint id: %s, high checkpoint id: %s, ignore base disk: %t)",
			getChangedBlockReq.DiskId,
			getChangedBlockReq.StartIndex,
			getChangedBlockReq.BlocksCount,
			getChangedBlockReq.LowCheckpointId,
			getChangedBlockReq.HighCheckpointId,
			getChangedBlockReq.IgnoreBaseDisk)
	}

	if createPlacementGroupReq, ok := req.(*protos.TCreatePlacementGroupRequest); ok {
		return fmt.Sprintf(
			" (group id: %s, placement strategy: %s, placement partition count: %d)",
			createPlacementGroupReq.GroupId,
			protos.EPlacementStrategy_name[int32(createPlacementGroupReq.PlacementStrategy)],
			createPlacementGroupReq.PlacementPartitionCount)
	}

	if destroyPlacementGroupReq, ok := req.(*protos.TDestroyPlacementGroupRequest); ok {
		return fmt.Sprintf(
			" (group id: %s)",
			destroyPlacementGroupReq.GroupId)
	}

	if alterPlacementGroupMembershipReq, ok := req.(*protos.TAlterPlacementGroupMembershipRequest); ok {
		return fmt.Sprintf(
			" (group id: %s, number of disks to add: %d, number of disks to remove: %d, placement partition index: %d)",
			alterPlacementGroupMembershipReq.GroupId,
			len(alterPlacementGroupMembershipReq.DisksToAdd),
			len(alterPlacementGroupMembershipReq.DisksToRemove),
			alterPlacementGroupMembershipReq.PlacementPartitionIndex)
	}

	return ""
}

File: cloud/blockstore/public/sdk/python/client/request.py
import datetime
import random
import sys

from concurrent import futures
from google.protobuf.message import Message

import cloud.blockstore.public.sdk.python.protos as protos


def _next_request_id() -> int:
    return random.randint(1, sys.maxsize)


def request_name(request: type[Message]) -> str:
    name = type(request).__name__
    if name.startswith("T") and name.endswith("Request"):
        # strip "T" prefix and "Request" suffix from type name
        return name[1:len(name) - 7]

    return name


def request_size(request: type[Message]) -> int:
    if isinstance(request, protos.TReadBlocksRequest):
        return request.BlocksCount

    if isinstance(request, protos.TWriteBlocksRequest):
        return len(request.Blocks.Buffers)

    if isinstance(request, protos.TZeroBlocksRequest):
        return request.BlocksCount

    return 0


def request_details(request: type[Message]) -> str:
    if isinstance(request, protos.TReadBlocksRequest):
        return " (offset: {}, count: {})".format(
            request.StartIndex,
            request.BlocksCount)

    if isinstance(request, protos.TWriteBlocksRequest):
        return " (offset: {}, count: {})".format(
            request.StartIndex,
            len(request.Blocks.Buffers))

    if isinstance(request, protos.TZeroBlocksRequest):
        return " (offset: {}, count: {})".format(
            request.StartIndex,
            request.BlocksCount)

    return ""


class _AsyncResponse(object):

    def __init__(
            self,
            future: futures.Future,
            request: type[Message],
            request_id: int,
            started: datetime.datetime):
        self.__future = future
        self.__request = request
        self.__request_id = request_id
        self.__started = started

    @property
    def future(self) -> futures.Future:
        return self.__future

    @property
    def request(self) -> type[Message]:
        return self.__request

    @property
    def request_id(self) -> int:
        return self.__request_id

    @property
    def started(self) -> datetime.datetime:
        return self.__started


class _RetryState(object):

    def __init__(
            self,
            request: type[Message],
            timeout: int,
            delay: float):
        self.__started = datetime.datetime.now()
        self.__request = request
        self.__timeout = timeout

        self.__delay = delay
        self.__retries = 0

        self.__response = futures.Future()

    @property
    def started(self) -> datetime.datetime:
        return self.__started

    @property
    def request(self) -> type[Message]:
        return self.__request

    @property
    def timeout(self) -> int:
        return self.__timeout

    @property
    def delay(self) -> float:
        return self.__delay

    @property
    def retries(self) -> int:
        return self.__retries

    @property
    def future(self) -> futures.Future:
        return self.__response

    def increment_delay(self, increment: float):
        self.__delay += increment

    def increment_retries(self):
        self.__retries += 1

    def set_result(self, result: type[Message]):
        self.__response.set_result(result)

    def set_exception(self, exception: type[BaseException]):
        self.__response.set_exception(exception)
